examples/pong.rs

```diff
@@ -0,0 +1,385 @@
+extern crate amethyst;
+
+use amethyst::engine::{Application, State, Trans};
+use amethyst::processors::rendering::{RenderingProcessor, Renderable, Light, Camera, Projection};
+use amethyst::context::Context;
+use amethyst::config::Element;
+use amethyst::ecs::{World, Join, VecStorage, Component, Processor, RunArg};
+use std::sync::{Mutex, Arc};
+
+struct Pong;
+
+struct Ball {
+    pub position: [f32; 2],
+    pub velocity: [f32; 2],
+    pub size: f32,
+}
+
+impl Ball {
+    pub fn new() -> Ball {
+        Ball {
+            position: [0.0, 0.0],
+            velocity: [-1.0, -1.0],
+            size: 1.0,
+        }
+    }
+}
+
+impl Component for Ball {
+    type Storage = VecStorage<Ball>;
+}
+
+enum Side {
+    Left,
+    Right,
+}
+
+struct Plank {
+    pub position: f32,
+    pub velocity: f32,
+    pub dimensions: [f32; 2],
+    pub side: Side,
+}
+
+impl Plank {
+    pub fn new(side: Side) -> Plank {
+        Plank {
+            position: 0.,
+            velocity: 1.,
+            dimensions: [1., 1.],
+            side: side,
+        }
+    }
+}
+
+impl Component for Plank {
+    type Storage = VecStorage<Plank>;
+}
+
+struct PongProcessor;
+
+unsafe impl Sync for PongProcessor {  }
+
+// If right_up == true then the right plank will move up,
+// other fields work the same way
+struct InputState {
+    pub right_up: bool,
+    pub right_down: bool,
+    pub left_up: bool,
+    pub left_down: bool,
+}
+
+impl InputState {
+    pub fn new() -> InputState {
+        InputState {
+            right_up: false,
+            right_down: false,
+            left_up: false,
+            left_down: false,
+        }
+    }
+}
+
+struct Score {
+    score_left: i32,
+    score_right: i32,
+}
+
+impl Score {
+    pub fn new() -> Score {
+        Score {
+            score_left: 0,
+            score_right: 0,
+        }
+    }
+}
+
+// Pong game processor
+impl Processor<Arc<Mutex<Context>>> for PongProcessor {
+    fn run(&mut self, arg: RunArg, context: Arc<Mutex<Context>>) {
+        use amethyst::context::event::{EngineEvent, Event, VirtualKeyCode, ElementState};
+        use std::ops::Deref;
+
+        // Get all needed component storages and resources
+        let context = context.lock().unwrap();
```

Comment:
I think it could work, but I haven't tried it yet. Probably it would make `Application` code less ugly, since it would allow removing all `context.lock.unwrap()` calls in `Application` impl and only `&mut World` would need to be passed to `State` methods instead of `&mut Context` and `&mut World`. Also it would simplify `Processor` implementation code, since you won't have to deal with `Arc<Mutex<Context>>` argument in `Processor::run` anymore.

But I think it would be better to put this change in a separate PR.